src/workerd/api/urlpattern-standard.h

```diff
@@ -0,0 +1,153 @@
+  };
+
+  // A structure providing matching patterns for individual components
+  // of a URL. When a URLPattern is created, or when a URLPattern is
+  // used to match or test against a URL, the input can be given as
+  // either a string or a URLPatternInit struct. If a string is given,
+  // it will be parsed to create a URLPatternInit. The URLPatternInit
+  // API is defined as part of the URLPattern specification.
+  struct URLPatternInit {
+#define V(_, name) jsg::Optional<kj::String> name;
+    URL_PATTERN_COMPONENTS(V)
+#undef V
+    jsg::Optional<kj::String> baseURL;
+
+    JSG_STRUCT(protocol, username, password, hostname, port, pathname, search, hash, baseURL);
+
+    ada::url_pattern_init toAdaType() const;
+  };
+
+  // A struct providing the URLPattern matching results for a single
+  // URL component. The URLPatternComponentResult is only ever used
+  // as a member attribute of a URLPatternResult struct. The
+  // URLPatternComponentResult API is defined as part of the URLPattern
+  // specification.
+  struct URLPatternComponentResult final {
+    jsg::JsString input;
+    jsg::JsObject groups;
+
+    JSG_STRUCT(input, groups);
+    JSG_STRUCT_TS_OVERRIDE({
+                  input: string;
+                  groups: Record<string, string>;
+                });
+  };
+
+  // A struct providing the URLPattern matching results for all
+  // components of a URL. The URLPatternResult API is defined as
+  // part of the URLPattern specification.
+  struct URLPatternResult final {
+    kj::Array<kj::OneOf<jsg::JsString, URLPatternInit>> inputs;
+#define V(_, name) URLPatternComponentResult name;
+    URL_PATTERN_COMPONENTS(V)
+#undef V
+
+    JSG_STRUCT(inputs, protocol, username, password, hostname, port, pathname, search, hash);
+  };
+
+  struct URLPatternOptions final {
+    jsg::Optional<bool> ignoreCase;
+
+    JSG_STRUCT(ignoreCase);
+
+    ada::url_pattern_options toAdaType() const;
+  };
+
+  explicit URLPattern(ada::url_pattern<URLPatternRegexEngine> i): inner(std::move(i)) {};
+
+  static jsg::Ref<URLPattern> constructor(jsg::Lock& js,
+      jsg::Optional<kj::OneOf<kj::String, URLPatternInit>> input,
+      jsg::Optional<kj::OneOf<kj::String, URLPatternOptions>> baseURL,
+      jsg::Optional<URLPatternOptions> patternOptions);
+
+  kj::Maybe<URLPatternResult> exec(jsg::Lock& js,
+      jsg::Optional<kj::OneOf<kj::String, URLPatternInit>> input,
+      jsg::Optional<kj::String> baseURL);
+
+  bool test(jsg::Optional<kj::OneOf<kj::String, URLPatternInit>> input,
+      jsg::Optional<kj::String> baseURL);
+
+  bool getHasRegExpGroups() const;
+
+#define V(name, _) kj::StringPtr get##name() const;
+  URL_PATTERN_COMPONENTS(V)
+#undef V
+
+  JSG_RESOURCE_TYPE(URLPattern) {
+#define V(Name, name) JSG_READONLY_PROTOTYPE_PROPERTY(name, get##Name);
+    URL_PATTERN_COMPONENTS(V)
+#undef V
+    JSG_READONLY_PROTOTYPE_PROPERTY(hasRegExpGroups, getHasRegExpGroups);
+    JSG_METHOD(test);
+    JSG_METHOD(exec);
+
+    JSG_TS_OVERRIDE({
```

Comment:
formatting feels ugly, maybe just inline to
`JSG_TS_OVERRIDE({ get hasRegExpGroups(): boolean; });` ?

Reply:
Do you even need a TS_OVERRIDE?

Reply:
I don't know why but we need it... The previous implementation doesn't have hasRegExpGroups getter. (which is a non-spec compliant bug)

Reply:
What's the generated types if you don't set the override?

Reply:
Basically it removes the getter.

Reply:
Does does rtti not support JSG_PROTOTYPE_READOLY_PROPERTY? Seems like we should solve that at the source!

Reply:
I've added this to my list. I'll take a look.